### ISSUE TITLE
Support different `ref` values for forward/backward directions on roads

### ISSUE THREAD

**Issue by danpat** [Project-OSRM/osrm-backend#4544]
OSRM doesn't (?) currently support different names in different directions on a single road.  We've lived with this because the situation is pretty rare.

However, the `ref` field is a bit different - roads frequently belong to different routes in opposite directions.  Fortunately, the really important roads that do this are usually divided roads, so we only *need* a single `ref` for them.

However, the work on adding [cardinal direction information](https://github.com/Project-OSRM/osrm-backend/pull/4524) has highlighted a problem: we can't currently support different `ref` names in different directions for single-way roads.

Many route relations use member roles like `forward` and `backward` to indicate that a road only belongs to a route in a certain direction.  We currently don't have a place to store this direction-specific information.

The TL;DR is that we need to support `result.forward_ref` and `result.backward_ref` so that we can properly handle different route names in different directions.

@MoKob I know you're looking at refactoring edge metadata storage at the moment, this seems like it's probably relevant to that.

/cc @deniskoronchik

**Comment by deniskoronchik** [comment]
Maybe it should be implemented as a complex storage of properties for each node? Why we talking about this problem in concept of `ref` value. I think OSRM require flexible storage that allows to store different attributes for and OSRM nodes.

/cc @danpat

**Comment by miccolis** [comment]
@danpat this case is crazy, a road is US-30 North one way and US-40 South the other? do you have an example handy of one of these roads?

**Comment by danpat** [comment]
@miccolis Actual different route names is fairly rare, but what's quite a bit more common is a road having a route _only in one direction_.  An example is this section in DC:

https://www.openstreetmap.org/way/47920789

It looks like US-29 actually splits for a while here, there is more of it on some roads to the south.

Also common is a mix of different interstate/state/county level routes, like those describe in the [wrong-way concurrencies](https://en.wikipedia.org/wiki/Concurrency_%28road%29#Wrong-way_concurrencies) wiki page, where some of them only apply in one direction.

**Comment by 1ec5** [comment]
> Also common is a mix of different interstate/state/county level routes, like those describe in the [wrong-way concurrencies](https://en.wikipedia.org/wiki/Concurrency_%28road%29#Wrong-way_concurrencies) wiki page, where some of them only apply in one direction.

Some statistics on the wrong-way concurrency phenomenon in https://github.com/Project-OSRM/osrm-text-instructions/issues/119#issuecomment-328053084.

**Comment by 1ec5** [comment]
`ref:forward` and `ref:backward` is also a common need in parts of Ohio and Indiana, where counties number their streets on a grid and county line roads can be a source of conflict.

Would this also make it possible to support roads that have a different `name` in each direction? I know of at least one county in Ohio that flips the names of major roads depending which direction you’re headed. For example, here’s [Cincinnati&harr;Columbus Road](http://www.openstreetmap.org/way/531362924) (US 42).

[<img width="292" alt="cin-col" src="https://user-images.githubusercontent.com/1231218/31366405-8d392fd0-ad25-11e7-9e74-b9562ccab280.png">](http://openstreetcam.org/details/522233/2635)
_Traveling northbound from towards Columbus, township signage constantly reminds me that I’m on Cincinnati Columbus Road._

[<img width="226" alt="col-cin" src="https://user-images.githubusercontent.com/1231218/31366444-bd97f60c-ad25-11e7-8223-4a3ab379b4d2.png">](http://openstreetcam.org/details/522233/2658)
_Intersection blade signs facing east say “Columbus Cincinnati Road”._

**Comment by TheMarex** [comment]
This was merged with https://github.com/Project-OSRM/osrm-backend/pull/4619
